### PR TITLE
GroupNormalization-21 updates

### PR DESCRIPTION
**Pull request by yuanyao-nv**
### Description

- Change the input shape of `scale` and `bias` from (G) to (C) to comply with the original paper
- Add `stash_type` attribute to specify the computation precision for the normalization step, similar to LayerNormalization.

### Motivation and Context

Closes https://github.com/onnx/onnx/issues/5466

### CODE REVIEW

**Comment by justinchuby** [comment]
Looks like we should disable `group_normalization_*` tests for onnxruntime. Do we know why the tests failed only on macs?

**Comment by yuanyao-nv** [comment]
> Looks like we should disable `group_normalization_*` tests for onnxruntime. Do we know why the tests failed only on macs?

Can you elaborate why they should be disabled? And how to disable them?
Looks like pytest cannot pick up these tests when I try locally.

**Comment by justinchuby** [comment]
Perhaps following this pattern: https://github.com/onnx/onnx/blob/c92866c6c82dba5647da5a8d67249c70c06824f1/onnx/test/test_backend_onnxruntime.py#L270

**Comment by justinchuby** [comment]
They need to be disabled because onnxruntime has not implemented support for this updated op. Although I remember seeing new op sets get filtered out automatically? @gramalingam what would you recommend?

File: onnx/backend/test/case/node/groupnormalization.py
Comment (gramalingam) [inline onnx/backend/test/case/node/groupnormalization.py:35]: Maybe using another symbolic name (for the value 4) would help clarify things to a reader?

Comment (gramalingam) [reply onnx/backend/test/case/node/groupnormalization.py:35]: Eg. use `C` is this is supposed to be C.

File: onnx/defs/nn/defs.cc
Comment (gramalingam) [inline onnx/defs/nn/defs.cc:2749]: May be clarify the allowed types (bfloat1 and float) here and maybe in documentation above also.

File: docs/Changelog.md
Comment (justinchuby) [inline docs/Changelog.md:24630]: PyTorch supports any number of dimensions after N x C (https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html). Should ONNX follow the same?

Comment (justinchuby) [inline docs/Changelog.md:24630]: Do we need to mention images cases here, or is the generic `(N x C x D1 x D2 ... Dn)` sufficient?

File: docs/Changelog.md
Comment (justinchuby) [inline docs/Changelog.md:24620]: PyTorch has the `num_channels` parameter (https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html) do you see it being supported?

Comment (justinchuby) [inline docs/Changelog.md:24620]: We would need `num_groups`, `num_channels` to be dynamic to support pytorch `native_group_norm` https://github.com/pytorch/pytorch/blob/1b506e74699afb33020c1a4cc6b05fa7fa263a9b/aten/src/ATen/native/native_functions.yaml#L2939C65-L2939C65

File: onnx/version_converter/convert.h
Comment (gramalingam) [inline onnx/version_converter/convert.h:614]: From our earlier discussion, I thought we had to do some transformation for the version-adapter (i.e., it is not a CompatibleAdapter). Has something changed or am I mistaken?

File: docs/Changelog.md
Comment (justinchuby) [inline docs/Changelog.md:24623]: Do we know the stash_type used in pytorch? Does it have a default type?

File: onnx/defs/nn/defs.cc
Comment (justinchuby) [inline onnx/defs/nn/defs.cc:2777]: I suggest using `OpSchema::all_float_types_ir4` or `OpSchema::all_float_types_ir9` for consistency

File: onnx/defs/nn/defs.cc
Comment (justinchuby) [inline onnx/defs/nn/defs.cc:2786]: Prefer lower case for variable names. Consider giving it a more meaningful name than "T"?

**Comment by gramalingam** [comment]
> They need to be disabled because onnxruntime has not implemented support for this updated op. Although I remember seeing new op sets get filtered out automatically? @gramalingam what would you recommend?

I am not sure about the test-infrastructure. However, looking at the error-message, I would check the function-definition to see if there is any issue with broadcasting (as implied by the error-message). There are one or two things about the test-infrastructure that puzzles me though (but that's a different concern): in principle it can run these tests since it is a function (so no kernel is required); however, I thought onnxruntime would reject opset-versions greater than what it handles (so unsure how it would accept this).

**Comment by gramalingam** [comment]
Ok, the failed test-case is actually one with opset 20 (not opset 21). At least that explains why the test-case ran. Need to look at the test-cases to see which opset versions are being targetted by the different test-cases

**Comment by gramalingam** [comment]
> Ok, the failed test-case is actually one with opset 20 (not opset 21). At least that explains why the test-case ran. Need to look at the test-cases to see which opset versions are being targetted by the different test-cases

I don't understand why the test-case has opset 20 ... IIUC, it should be 21, unless the test-case explicitly specifies a target opset (which doesn't happen in this PR).

File: docs/Changelog.md
Comment (yuanyao-nv) [reply docs/Changelog.md:24620]: In the link you provided, it seems like  `group` is an int?
`C` is a SymInt, just like in the ONNX spec C is inferred from the input shape.

Comment (yuanyao-nv) [reply docs/Changelog.md:24620]: `num_channels` can be inferred from the input shape already - that's why it wasn't included here.
We can consider including it for better conformity with torch. Do you have a preference? cc: @gramalingam

File: docs/Changelog.md
Comment (yuanyao-nv) [reply docs/Changelog.md:24630]: It's mentioned a couple of sentences before this: "Dimensions for image cases are `(N x C x H x W)`"

File: docs/Changelog.md
Comment (yuanyao-nv) [reply docs/Changelog.md:24623]: PyTorch does not have a stash_type attribute. This attribute is useful for the inference setting, just like in the case of LayerNormalization.

Comment (justinchuby) [reply docs/Changelog.md:24623]: Right. But do you have an implicit default? Do we know what PyTorch does on the background?

File: onnx/defs/nn/defs.cc
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: updated.

File: docs/Changelog.md
Comment (justinchuby) [reply docs/Changelog.md:24620]: I think you are right. Does it mean all N, C and HxW are inferred?

Comment (justinchuby) [reply docs/Changelog.md:24620]: If it is inferred I wouldn't bother adding that, thanks!

File: docs/Changelog.md
Comment (justinchuby) [reply docs/Changelog.md:24623]: So should we allow an option, for example `-1`, to mirror PyTorch's behavior?

File: docs/Changelog.md
Comment (gramalingam) [reply docs/Changelog.md:24620]: Agree. I don't think we should add it unless there is a good reason. May be Pytorch needs it to allocate the scale/bias tensors that are learnt via training? For us, they (scale/bias) are extra inputs.

Comment (yuanyao-nv) [reply docs/Changelog.md:24620]: Yes they all can be inferred from input shape.

File: docs/Changelog.md
Comment (yuanyao-nv) [reply docs/Changelog.md:24623]: I think it's a matter of taste. From a utility point of view it doesn't add anything more, since the user can simply set stash_type to match input type.

And if we do decide to make this change we should probably update the behavior of LayerNorm as well.

File: onnx/version_converter/convert.h
Comment (yuanyao-nv) [reply onnx/version_converter/convert.h:614]: @gramalingam Yes I need to change this part.

If we use an adapter, the logic should look like this:
```
    -- Shape<start=1,end=2>() -- Div(Shape_out, num_groups) 
  /                                             \
--  Reshape(,[-1,1]) ----------------- Expand(Reshape_out,[1,Div_out]) -- Reshape(, [-1])
```
There are a few pieces required here whose syntax I haven't quite figured out here:
- how to set attribute of the Shape node
- Turn `num_groups` into an initializer
- Constructing something like `[1, Div_out]`. Is it necessary to use a Constant node for `[1]` and then a Concat node to get `[1,Div_out]`?
I'm trying to see if I can find some similar examples in other adapters.

I think another option you mentioned previously was simply for this version conversion to fail. Is that an acceptable option? What are these adapters here used for: will they be called by third-party conversion tools? If so, I wonder if this will be useful since there will probably be no model out there that currently uses the opset18 definition.

File: docs/Changelog.md
Comment (justinchuby) [reply docs/Changelog.md:24630]: Right. I am just thinking the image case can be removed. But I don’t have a strong opinion.

File: onnx/defs/nn/defs.cc
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc:2749]: On second thought, I think we should relax the restriction on stash_type to allow any dtype that the input can take. The motivation for having a stash_type is so that if necessary the user can choose to use a higher precision than input dtype for stage 1 computations. But I don't think the spec itself should prohibit user from using any desired floating point type.

In the LayerNorm case, the spec also similarly doesn't put any restrictions, but the function definition itself asserts stash_type to be float or bfloat16, which IMO is overly restrictive.

File: docs/Changelog.md
Comment (justinchuby) [reply docs/Changelog.md:24623]: When we create ONNX functions for pytorch operators, it would be good to be able to follow pytorch behavior. Otherwise the ONNX functions will need to be specialized for each invocation. @gramalingam

File: onnx/defs/nn/defs.cc
Comment (gramalingam) [reply onnx/defs/nn/defs.cc:2749]: The proposed generalization seems ok to me. Is that float/blfoat16/float16/double ? I suppose we actually would need this for backward compatibility (when doing version-conversion from older version)?

File: docs/Changelog.md
Comment (gramalingam) [reply docs/Changelog.md:24623]: I see @yuanyao-nv mentions in another comment-thread that it would be better to allow all input types as stash type. If/when we do that, it seems more natural to say that the default value of stash type is the same as input type? This would also be more consistent with the older GroupNorm op (otherwise, the version-converter would need to add the correct stash-type attribute).

I do see the point of making this consistent with LayerNorm. Perhaps that could be done as part of the other proposal to unify the various normalization functions?

File: onnx/defs/nn/defs.cc
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc:2749]: @gramalingam Yes, it would be the same set of dtypes as input, so float/blfoat16/float16/double.
And yes, backward compatibility is a good point too.

File: docs/Changelog.md
Comment (yuanyao-nv) [reply docs/Changelog.md:24623]: @gramalingam Are you aware of any reason LayerNorm didn't make the default `stash_type` match the input type? I wonder if there are other considerations like kernel optimization which requires the stage 1 precision to be known statically. If we make it match the input dtype then it could become a dynamically known precision.

On the other hand, IMO backward compatibility is a nice-to-have but is also a lesser concern since I'm not aware of models that use GroupNorm-18 yet due to its over-restrictiveness.

**Comment by justinchuby** [comment]
Mac issues should be fixed with https://github.com/onnx/onnx/pull/5814

**Review by justinchuby — APPROVED** [review]
LGTM with existing comments resolved. Long awaited improvement 🚀

**Comment by justinchuby** [comment]
> Perhaps following this pattern:
> 
> https://github.com/onnx/onnx/blob/c92866c6c82dba5647da5a8d67249c70c06824f1/onnx/test/test_backend_onnxruntime.py#L270

These tests still needs to be skipped

```
FAILED onnx/test/test_backend_onnxruntime.py::OnnxBackendNodeModelTest::test_group_normalization_epsilon_cpu - RuntimeError: Unable to create inference session. Model is:
FAILED onnx/test/test_backend_onnxruntime.py::OnnxBackendNodeModelTest::test_group_normalization_epsilon_expanded_cpu - RuntimeError: Unable to create inference session. Model is:
FAILED onnx/test/test_backend_onnxruntime.py::OnnxBackendNodeModelTest::test_group_normalization_example_cpu - RuntimeError: Unable to create inference session. Model is:
FAILED onnx/test/test_backend_onnxruntime.py::OnnxBackendNodeModelTest::test_group_normalization_example_expanded_cpu - RuntimeError: Unable to create inference session. Model is:
```

File: docs/Changelog.md
Comment (yuanyao-nv) [reply docs/Changelog.md:24630]: In keeping with the convention of a few other ops, I'd say let's keep it here?

File: onnx/version_converter/convert.h
Comment (yuanyao-nv) [reply onnx/version_converter/convert.h]: @gramalingam I added a version adapter based on the logic above. I'm not sure if it's sufficiently covered in the integration tests. Do you have a recommended way of verifying it?

File: onnx/version_converter/convert.h
Comment (justinchuby) [inline onnx/version_converter/convert.h:615]: It would be amazing to have a down convert adapter implemented as well. It doesn’t have to use the old group norm operator if needed. This is going to be important when we move to use the latest opset while maintaining the ability to convert back to older opsets. 


FWIW this is now we implement group norm right now: https://github.com/microsoft/onnxscript/blob/172a81745c10467e04d9b4fdfa7829ce3321b6ce/onnxscript/function_libs/torch_lib/ops/core.py#L5850

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (justinchuby) [inline onnx/version_converter/adapters/group_normalization_20_21.h:98]: Need to #include <memory> ?

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (justinchuby) [inline onnx/version_converter/adapters/group_normalization_20_21.h:65]: Call it constant_one etc? Same for constant1, constant3

File: onnx/version_converter/convert.h
Comment (yuanyao-nv) [reply onnx/version_converter/convert.h:615]: You mean converting from opset 21 to 20? I don't think it's well defined how to convert a large tensor to a small tensor for scale and bias.

As I mentioned in the other thread, even the utility of 20->21 seems a bit dubious to me since I don't think any model should use the opset 20 definition - and I'm not aware of any.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h:98]: Could you elaborate what you mean here?

File: onnx/version_converter/convert.h
Comment (justinchuby) [reply onnx/version_converter/convert.h:615]: Could you elaborate on what you mean by convert a large tensor to a small tensor for scale and bias?

I agree that 20->21 can be less useful because very few are currently using the op. However when converters are adopting the new operator in opset 21, there will be needs for converting to lower opsets when users are trying to use runtimes that have not caught up with support.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (justinchuby) [reply onnx/version_converter/adapters/group_normalization_20_21.h:98]: See the code comment in the code tab: [cpplint] reported by reviewdog 🐶
 Add #include <memory> for shared_ptr<>  [build/include_what_you_use] [4]

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h]: Updated these variable names.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h:100]: Thanks, added.

File: onnx/version_converter/convert.h
Comment (yuanyao-nv) [reply onnx/version_converter/convert.h:615]: The difference between the two opsets is that opset20 has scale/bias shape as (num_groups,) and opset 21 has it as (C,).

The opset20 definition is in fact incorrect since it doesn't match the original paper. To go from 20 to 21, we can go through the sequence of operations defined in the adapter so that, e.g., scale=[1,2,3] becomes scale=[1,1,2,2,3,3] when C=6 and num_grps=3. However, it does not work in the other direction: suppose scale=[1,2,3,4,5,6] in opset21, we don't know how to convert it to a length-3 tensor.

Comment (justinchuby) [reply onnx/version_converter/convert.h:615]: Sounds like we don’t need to use the same operator when converting back, as long as the model is correct and accurate. I think we can simply reuse the function definition (defined in this or) or refer to the implementation above (which uses instance norm) to create the opset20 adapter. Aka we don’t need to use GroupNorm. Does this address the concern?

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (justinchuby) [inline onnx/version_converter/adapters/group_normalization_20_21.h:36]: Nit: reminder to avoid abbreviations and to spell words out, even though it can be clear that grp is group, it is unnecessarily abbreviated and adds cognitive load to readers.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (gramalingam) [inline onnx/version_converter/adapters/group_normalization_20_21.h:33]: Shouldn't the input-index used be `0` instead of `input_id`?

Comment (gramalingam) [reply onnx/version_converter/adapters/group_normalization_20_21.h:33]: Further, the Shape and Div part is common for both scale and bias. So, it could be factored out and done once (instead of repeating it for both), right?

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (gramalingam) [inline onnx/version_converter/adapters/group_normalization_20_21.h:30]: minor nit: May be naming this `C` instead of `shape` would help clarify what this is intended to be.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (gramalingam) [inline onnx/version_converter/adapters/group_normalization_20_21.h:58]: Change `1` to `input_id` ?

File: docs/Changelog.md
Comment (gramalingam) [reply docs/Changelog.md:24623]: > I wonder if there are other considerations like kernel optimization which requires the stage 1 precision to be known statically.

I don't think that is a consideration. In principle, it is still known statically since all types are known statically. I am not sure why LayerNorm had this default. There may have been no conscious reason.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (gramalingam) [reply onnx/version_converter/adapters/group_normalization_20_21.h:33]: The reshape shapes are also common to both scale and bias, I guess.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h:30]: It's actually not `C` here yet. The adapter will change it to `C` eventually.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h:33]: It should be `input_id` since this is a helper function that's already factored out to be applied to both scale and bias.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h]: Thanks, Good catch. Updated.

File: onnx/version_converter/convert.h
Comment (justinchuby) [reply onnx/version_converter/convert.h:615]: Happy to unblock this pr and follow up as a separate task

Comment (justinchuby) [reply onnx/version_converter/convert.h:615]: https://github.com/onnx/onnx/issues/5843

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (gramalingam) [reply onnx/version_converter/adapters/group_normalization_20_21.h:33]: I am confused. Based on the description elsewhere, my thinking is this:

Input model (older opset): shapes of operands are:
X (input 0): [ N, C, ...]
scale (input 1): [num_groups]
bias (input 2): [num_groups]

We want to change scale and bias to have shape [C].

Is this understanding correct? Or, wrong?

Comment (gramalingam) [reply onnx/version_converter/adapters/group_normalization_20_21.h:33]: And, I think, the code is trying to calculate the value of `C / num_groups`. The value of `C` has to come from the shape of input 0. What am I missing?

Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h]: Ah sorry I was misled by my diagram. You're right - the Shape operator should get the input from the 0th input and the Reshape operator gets the input from the 1st and 2nd inputs. I've also corrected the diagram.

I also refactored some common code for scale and bias into the transform_input() function and left the constant nodes outside to avoid duplication. Please let me know if this is what you meant.

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h]: Scratch the above. I updated it to `C`.

**Comment by justinchuby** [comment]
Could you resolve conflicts? Thanks!

File: onnx/version_converter/adapters/group_normalization_20_21.h
Comment (yuanyao-nv) [reply onnx/version_converter/adapters/group_normalization_20_21.h]: @gramalingam Let me know if these changes look good to you. Thanks!

**Comment by justinchuby** [comment]
Should be ready to merge after the merge conflicts are fixed.

File: onnx/defs/nn/defs.cc
Comment (xadupre) [inline onnx/defs/nn/defs.cc:2732]: depend

File: docs/Changelog.md
Comment (justinchuby) [inline docs/Changelog.md:24873]:
```suggestion
  original type of `X`. The second stage does not depend on `stash_type`.
```

File: onnx/defs/nn/defs.cc
Comment (justinchuby) [inline onnx/defs/nn/defs.cc:2732]:
```suggestion
original type of `X`. The second stage does not depend on `stash_type`.
```

File: docs/Operators.md
Comment (justinchuby) [inline docs/Operators.md:11410]:
```suggestion
  original type of `X`. The second stage does not depend on `stash_type`.
```

**Review by justinchuby — APPROVED** [review]

**Review by xadupre — APPROVED** [review]

**Pull request merged**
